package/MDAnalysis/topology/ITPParser.py

```diff
@@ -111,29 +148,45 @@ def parse(self, **kwargs):
         charges = []
         masses = []
 
-        atom_lists = [ids, types, resids, resnames, names, chargegroups, charges, masses]
+        atom_lists = [self.ids, types, resids, resnames, names, chargegroups, charges, masses]
 
-        bonds = []
-        bondtypes = []
-        angles = []
-        angletypes = []
-        dihedrals = []
-        dihedraltypes = []
-        impropers = []
-        impropertypes = []
+        bonds = defaultdict(list)
+        angles = defaultdict(list)
+        dihedrals = defaultdict(list)
+        impropers = defaultdict(list)
 
         segid = 'SYSTEM'
         molnum = 1
 
         section = None
+        if_queue = []
 
         # Open and check itp validity
-        with openany(self.filename) as itpfile:
-            for line in itpfile:
+        with openany(self.filename) as self.itpfile:
+            for line in self.itpfile:
                 line = line.split(';')[0].strip()  # ; is for comments
                 if not line:  # Skip line if empty
                     continue
-                if line.startswith('#'): #ignore include, ifdefs, etc
+
+                if line.startswith('#ifdef'):
```

Comment:
Maybe being pedantic, are nested `ifdef`s allowed in itp files?  If so I think maybe this approach won't work?

Reply:
I am almost sure they are allowed.

Reply:
In which case you need some sort of if stack thing like it looks like was planned.

Reply:
I think I did something similar to @jbarnoud's ``_skip_lines_until_endif()`` thing, although without the IOError catching. There's a test case for nested ifs.